### PR TITLE
Fix image origin consistently.

### PR DESCRIPTION
**Pull request by mapazarr**
Fix image origin for WCS coordinate transformations in `wcs_world2pix` and `wcs_pix2world` method.
This pull request addresses issue #251.

### CODE REVIEW

File: gammapy/image/utils.py [hunk 1]
```diff
@@ -387,7 +387,7 @@ def coordinates(image, world=True, lon_sym=True, radians=False):
         y, x = np.indices(image.data.shape, dtype='int32') + 1
 
     wcs = WCS(image.header)
-    lon, lat = wcs.wcs_pix2world(x, y, 1)
+    lon, lat = wcs.wcs_pix2world(x, y, 0)
```
Comment (cdeil) [inline gammapy/image/utils.py]: I think that `origin=1` was correct here, because above we use `np.indices(...) + 1`.
Probably the `+1` above should be removed and `origin=0` used here, to make it more clear that we use the normal Astropy pixel coordinate convention.

But the first thing to do is to write tests for this central function in Gammapy.
E.g. one could write images to FITS and check results by zomming in to the pixel in the lower-left corner in ds9.

@mapazarr – Let me know if this is something you want to continue or if I should take over. I'd guess that adding tests for this (and maybe fixing bugs in this function or others that are discovered) can easily become quite a bit of work.

Comment (mapazarr) [inline gammapy/image/utils.py]: Ok, you can then take over. If you modify the `coordinates` function in `gammapy/image/utils.py` in the way you mention, you may want to revert the changes in the hard coded references to the number of pixels in `test_image_lon_profile` from ` gammapy/image/tests/test_profile.py `.

Comment (cdeil) [inline gammapy/image/utils.py]: I don't know when I'll get to this. 
For now I've put an entry describing what our pixel coordinate convention is in the docs:
https://gammapy.readthedocs.org/en/latest/development/index.html#pixel-coordinate-convention

**Comment by mapazarr** [comment]
I reverted the changes to the `image.utils.coordinates` function and the `test_image_lon_profile` test.
I also reviewed the use of the correct origin in other functions:

* `SkyCoord.skycoord_to_pixel`
* `SkyCoord.pixel_to_skycoord`
* `SkyCoord.from_pixel`
* `SkyCoord.from_pixel`

They are not used in the code.

**Comment by cdeil** [comment]
These methods on `SkyCoord` were only added recently, that's why they aren't used.

They probably should be used in a lot of places, but I don't have time to review the codebase.
So @mapazarr, my suggestion would be that whenever you think something is wrong or can be improved, you ask on the Gammapy mailing list or if it's clear, directly make an issue or pull request on Github.

In this case the only remaining change is in `gammapy/morphology/overlap.py`, and that code should probably be removed completely, right, @adonath ?

**Comment by mapazarr** [comment]
Well, there is still the issue with `image.utils.coordinates`: it still uses "1" as a reference instead of "0". You mentioned that you wanted to do some tests before correcting it. But the other changes are ready to be merged.

**Comment by cdeil** [comment]
I had another look at the `coordinates` function.
It currently does this:
```python
    # Create arrays of pixel coordinates
    if not world:
        y, x = np.indices(image.shape, dtype='int32') + 1
        return x, y
    else:
        y, x = np.indices(image.data.shape, dtype='int32') + 1

    wcs = WCS(image.header)
    lon, lat = wcs.wcs_pix2world(x, y, 1)
```

I think it should do this:
```python
    # Create arrays of pixel coordinates
    y, x = np.indices(image.shape, dtype='int32')

    if not world:        
        return x, y

    wcs = WCS(image.header)
    lon, lat = wcs.wcs_pix2world(x, y, origin=0)
```

Note that this would give the same output when `world=True`, because `1` was changed to `0` in two places.

What I didn't realise was that it will give an output shifted by 1 when `world=False`.

While this PR is still open, could you please make this change and if you have time, check (e.g. using the Github full-text search) if this function is ever called with `world=False` and if yes, we'll look if there's a bug there that needs to be fixed.

File: gammapy/morphology/overlap.py [hunk 2]
```diff
@@ -175,7 +175,7 @@ def get_containment_mask(glon_pos, glat_pos, r_containment, shape):
     hdulist = fits.open('../counts.fits')
     w = WCS(hdulist[0].header)
     y, x = np.indices(shape)
-    glon, glat = w.wcs_pix2world(x, y, 1)
+    glon, glat = w.wcs_pix2world(x, y, 0)
```
Comment (cdeil) [inline gammapy/morphology/overlap.py]: Could you please everywhere change this to explicitly pass `origin=0` via a keyword argument (instead of just `0` as third positional argument)?
I think this makes for more readable code.

**Comment by cdeil** [comment]
At the moment there's three commits in this PR for a change that should be a single commit.

Could you please use this as an opportunity to practice your git-fu and squash those commits by saying `git rebase master --interactive` followed by `git push -f <remote_name> <branch_name>`, i.e. force-push the new commit to Github?
(See http://astropy.readthedocs.org/en/latest/development/workflow/additional_git_topics.html?highlight=squash#rewriting-commit-history)

File: gammapy/morphology/overlap.py (hunk 2 above)
Comment (mapazarr) [inline gammapy/morphology/overlap.py]: Fixed.

**Comment by mapazarr** [comment]
I modified the function `image.utils.coordinates` as proposed. There is no explicit call to it with 'world=False`, so I guess we are in the safe side modifying the function.

**Comment by cdeil** [comment]
I don't see an update here. Did you want to do something else or can you push now?

**Comment by mapazarr** [comment]
I didn't push the changes because of the problem with the "origin" keyword. Since it's not recognized, I would need to define the variables: should I do that?

File: gammapy/morphology/overlap.py (hunk 2 above)
Comment (cdeil) [inline gammapy/morphology/overlap.py]: Ah, sorry, I missed this comment.

I thought `origin` was an explicit argument of `WCS.wcs_world2pix`, but now I see that the function signature is `wcs_pix2world(*args, **kwargs)` and if `origin` is passed as a keyword argument it's not handled by the function. I would say that's a bad API, but I don't think it makes sense to complain about this now that it's been like this for years.

So I would say leave as is then, just pass `0` as third parameter.

The alternatives would be to every time write
```python
origin = 0 # Astropy pixel coordinate convention
wcs.wcs_world2pix(x, y, origin)
```
or
```python
# Note: third parameter is `origin=0`, following the Astropy pixel convention
wcs.wcs_world2pix(x, y, 0)
```
which would be fine with me as well, but is a bit tedious.
@mapazarr – You decide.

**Comment by mapazarr** [comment]
Ok I rebased and forced the push.
Now everything should work. I also noticed that the test for `background.fill_acceptance_image` uses `image.utils.coordinates` with `world=False`. I adapted the test correspondingly to use the new convention.
https://github.com/gammapy/gammapy/blob/master/gammapy/background/tests/test_fov.py

File: gammapy/background/tests/test_fov.py [hunk 3]
```diff
@@ -66,13 +66,13 @@ def test_fill_acceptance_image():
...
+    coord = pixel_to_skycoord(xpix_coord_grid, ypix_coord_grid, w, origin=0)
     pix_off = coord.separation(center)
 
-    # x axis defined in the array positions [y_center_pix - 1,:]
-    # only interested in semi axis, so [y_center_pix - 1, x_center_pix - 1:]
-    ix_min = int(round(x_center_pix - 1))
-    iy = int(round(y_center_pix - 1))
+    # x axis defined in the array positions [y_center_pix,:]
+    # only interested in semi axis, so [y_center_pix, x_center_pix:]
+    ix_min = int(round(x_center_pix))
```
Comment (cdeil) [inline gammapy/background/tests/test_fov.py]: This fails on Python 3:
https://travis-ci.org/gammapy/gammapy/jobs/57998355#L1045

File: gammapy/data/spectral_cube.py [hunk 4]
```diff
@@ -157,7 +157,8 @@ def world2pix(self, lon, lat, energy, combine=False):
         """
         lon = lon.to('deg').value
         lat = lat.to('deg').value
-        x, y, _ = self.wcs.wcs_world2pix(lon, lat, 0, 0)
+        origin=0  # convention for gammapy
```
Comment (cdeil) [inline gammapy/data/spectral_cube.py]: PEP 8 says to put spaces here, please do that everywhere you put this line:
```python
origin = 0
```

**Comment by mapazarr** [comment]
@cdeil & @adonath:
I don't understand why travis complains about the test. The error is:
`TypeError: type numpy.ndarray doesn't define __round__ method`
But it had worked in the past (otherwise the test wouldn't be merged). I changed:
`ix_min = int(round(x_center_pix - 1))`
for
`ix_min = int(round(x_center_pix))`.
Why did it work before and it's not working now?
(BTW, in my local machine the test runs without problems).

File: gammapy/data/spectral_cube.py (hunk 4 above)
Comment (mapazarr) [inline gammapy/data/spectral_cube.py]: Fixed.

File: gammapy/background/tests/test_fov.py (hunk 3 above)
Comment (mapazarr) [inline gammapy/background/tests/test_fov.py]: Fixed. (Still not sure why it failed, see my comment below).

**Comment by mapazarr** [comment]
Ok I removed the polemic function and rebased again. Now it's working.

**Comment by cdeil** [comment]
@mapazarr – Merging this now, thanks!

About why your test stopped working with Python 3 in this PR, probably some minor change that lead to `x_center_pix` now being a Numpy array, whereas it was a scalar before.
I don't have time to look in detail, and you don't have to either ... it works now, so it's OK.

For the future: In case you don't have it already, create both a Python 2.7 and 3.4 conda environment locally that you can regularly use for testing. I have Python 2 and 3 from Macports and I see this:
```python
$ python2.7
>>> import numpy as np
>>> round(np.array(1.2))
1.0
```
```python
$ python3.4
>>> import numpy as np
>>> round(np.array(1.2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: type numpy.ndarray doesn't define __round__ method
>>> np.round(np.array(1.2))
1.0
```
Sometimes it's even necessary to make a conda environment with a specific version of numpy or scipy or matplotlib or Astropy to reproduce and find a workaround for some weird issue, but just testing on Python 2 and 3 reveals and lets you debug > 90% of the issues.
